### ISSUE TITLE
Horizontal white lines/gaps when tiling image glyphs

### ISSUE THREAD

**Issue by akloster** [bokeh/bokeh#8106]
I'm trying to tile image glyphs. According to the documentation, "dilate=True" is supposed to make image glyphs seamless. In my example, I see horizontal white stripes at certain zoom stages.

bokeh 0.12.16
IPython 6.4.0
Python 3.6.5
Notebook 5.5.0
Chrome 67.0.3396.99

```
import numpy as np
import scipy as sp
import scipy.misc
import numpy as np
from bokeh.plotting import figure, show, output_notebook
from bokeh.models import Color
output_notebook()

def to_rgba():
    N = 100
    out = np.zeros((N,N), dtype=np.uint32)
    view = out.view(dtype=np.uint8).reshape((N, N, 4))    
    m,n = N,N
    view[:m, :n, 0] = 0
    view[:m, :n, 1] = 0
    view[:m, :n, 2] = 0
    view[:m, :n, 3] = 255
    return out
images = []
pos_x = []
pos_y = []
for i in range(4):
    for j in range(4):
        img = to_rgba()
        images.append(img)
        pos_x.append(i*100)
        pos_y.append(j*100)
p = figure(x_range=(0,400), y_range=(0,400), output_backend="webgl", plot_width=400, plot_height=400)
p.xgrid.grid_line_color = None
p.ygrid.grid_line_color = None
p.image_rgba(image=images,
        x=pos_x, y=pos_y, dw=100, dh=100, dilate=True,
        global_alpha=1,
       )
show(p)
```

**Comment by bryevdv** [comment]
I'm not quite sure what to do about this, setting `dilate` causes the ceiling to always be used for the pixel width and height calculations. We could just add some arbitrary extra amount, on top of that, but that does not seem correct.

**Comment by akloster** [comment]
I also don't know. I've checked that the rect glyph with dilate=True tiles seamlessly. I wanted to make a plot where the tiles are computed from a dask volume and updated as they arrive.

**Comment by bryevdv** [comment]
Well, rects normally also have a line width, which straddles the border, so perhaps adding a *half* pixel would be reasonable for dilate.

**Comment by akloster** [comment]
I made another example with a blue column/row at 0,0 and a red column/row at 99,99
![image](https://user-images.githubusercontent.com/171319/43190533-90d459aa-8ff9-11e8-92d1-1b156f20023e.png)

In that picture the blue horizontal bar is 22 pixels high, the red one 23. The white line seems to be in the blue region.

```import numpy as np
import scipy as sp
import scipy.misc
import numpy as np
from bokeh.plotting import figure, show, output_notebook
from bokeh.models import Color
output_notebook()

def to_rgba():
    N = 100
    out = np.zeros((N,N), dtype=np.uint32)
    view = out.view(dtype=np.uint8).reshape((N, N, 4))    
    view[N-1, :,0] = 255
    view[0,:,2] = 255
    view[:, N-1:,0] = 255
    view[:,0,2] = 255
    view[:, :, 3] = 255
    return out


images = []
pos_x = []
pos_y = []
for i in range(4):
    for j in range(4):
        img = to_rgba()
        images.append(img)
        pos_x.append(i*100)
        pos_y.append(j*100)
p = figure(x_range=(0,400), y_range=(0,400), output_backend="webgl", plot_width=400, plot_height=400)
p.xgrid.grid_line_color = None
p.ygrid.grid_line_color = None
p.image_rgba(image=images,
        x=pos_x, y=pos_y, dw=100, dh=100, dilate=True,
        global_alpha=1,
       )
show(p)
```

**Comment by bryevdv** [comment]
@akloster pointed this out on the gitter channel:

https://aapm.onlinelibrary.wiley.com/doi/abs/10.1002/acm2.12401

TLDR; apparently is common/recommended practice when displaying tiles to create tiles that have some level of overlap at the borders, specifically to contend with situations such as this. As Bokeh itself is not involved in the *creation* of tiles, there is no solution that can be implemented at the Bokeh level. The current placement reflects a straightforward linear mapping of data space dimensions to pixel space via a simple linear transformation, and an option to always "round out" the result. I don't see any other appropriate things we could do.
